fem/hyperbolic.hpp

```diff
@@ -0,0 +1,446 @@
+#ifndef MFEM_THREAD_SAFE
+   mutable DenseMatrix flux;
+#endif
+};
+
+
+/**
+ * @brief Abstract class for numerical flux for a system of hyperbolic conservation laws
+ * on a face with states, fluxes and characteristic speed
+ *
+ */
+class RiemannSolver
+{
+public:
+   RiemannSolver(const FluxFunction &fluxFunction):fluxFunction(fluxFunction)
+   {
+   }
+   /**
+    * @brief Evaluates numerical flux for given states and fluxes. Must be
+    * overloaded in a derived class
+    *
+    * @param[in] state1 state value at a point from the first element
+    * (num_equations)
+    * @param[in] state2 state value at a point from the second element
+    * (num_equations)
+    * @param[in] nor scaled normal vector, @see mfem::CalcOrtho (dim)
+    * @param[in] Tr face information
+    * @param[out] flux numerical flux (num_equations)
+    */
+   virtual double Eval(const Vector &state1, const Vector &state2,
+                       const Vector &nor, FaceElementTransformations &Tr,
+                       Vector &flux) const = 0;
+   virtual ~RiemannSolver() = default;
+protected:
+   const FluxFunction &fluxFunction;
+};
+
+/**
+ * @brief Abstract hyperbolic form integrator, (F(u, x), ∇v) and (F̂(u±, x, n))
+ *
+ */
+class HyperbolicFormIntegrator : public NonlinearFormIntegrator
+{
+private:
+   // The maximum characterstic speed, updated during element/face vector assembly
+   double max_char_speed;
+   const FluxFunction &fluxFunction;
+   const int IntOrderOffset;  // 2*p + IntOrderOffset will be used for quadrature
+   const RiemannSolver &rsolver;    // Numerical flux that maps F(u±,x) to hat(F)
+#ifndef MFEM_THREAD_SAFE
+   // Local storages for element integration
+   Vector shape;              // shape function value at an integration point
+   Vector state;              // state value at an integration point
+   DenseMatrix flux;          // flux value at an integration point
+   DenseMatrix dshape;  // derivative of shape function at an integration point
+
+   Vector shape1;  // shape function value at an integration point - first elem
+   Vector shape2;  // shape function value at an integration point - second elem
+   Vector state1;  // state value at an integration point - first elem
+   Vector state2;  // state value at an integration point - second elem
+   Vector fluxN1;  // flux dot n value at an integration point - first elem
+   Vector fluxN2;  // flux dot n value at an integration point - second elem
+   Vector nor;     // normal vector, @see CalcOrtho
+   Vector fluxN;   // hat(F)(u,x)
+#endif
+
+public:
+   const int num_equations;  // the number of equations
+   /**
+    * @brief Construct a new Hyperbolic Form Integrator object
+    *
+    * @param[in] fluxFunction governing equation
+    * @param[in] rsolver numerical flux
+    * @param[in] IntOrderOffset 2*p+IntOrderOffset order Gaussian quadrature
+    * will be used
+    */
+   HyperbolicFormIntegrator(
+      const FluxFunction &fluxFunction, const RiemannSolver &rsolver,
```

Comment:
Does this `fluxFunction` reference have to match the once used in constructing the object referenced by `rsolver`?

Reply:
I think this is a very good point. This can prevent user-error. Thank you for your suggestion. I changed `RiemannSolver` as you suggested. I also added `RiemannSolver::IsCompatible(const FluxFunction &f)`. This will be called in the constructor of `HyperbolicFormIntegrator` to prevent user from using a system-specific Riemann Solver with a different system.

Reply:
Ah ok, I was wondering how you were thinking about including exact Riemann solvers vs approximate Riemann solvers. If you just wanted to support approximate Riemann solvers, then there's no need to check for compatibility? If want to support both solvers with the same `RiemannSolver` interface, then maybe it is better to have the derived `RiemannSolver` class provide the const reference... i.e., add `const FluxFunction& RiemannSolver::GetFluxFunction()` that the `HyperbolicFormIntegrator` class can use instead of storing a separate `FluxFunction` reference (constructor of `HyperbolicFormIntegrator` to just take the `RiemannSolver` reference instead of also taking the `FluxFunction` reference). What do you think?

Reply:
If considering a use case for `RiemannSolver` and `FluxFunction` beyond feeding into `HyperbolicFormIntegrator`, it might be more appropriate to check compatibility within the constructor of a derived class of `RiemannSolver`. However, continually invoking `RiemannSolver::GetFluxFunction()` during the element assembly of `HyperbolicFormIntegrator` may not be the most natural approach.

An alternative solution is as follows:
- Both `HyperbolicFormIntegrator` and `RiemannSolver` hold a const reference to `FluxFunction`. Compatibility check can be done in the constructor of the derived class of `RiemannSolver`.
- `HyperbolicFormIntegrator` only takes `RiemannSolver` in the constructor and stores the const reference to `FluxFunction` by calling `RiemannSolver::GetFluxFunction()`.
- Remove `FluxFunction` from the input list of `RiemannSolver::Eval`.

This approach ensures error-proofing without requiring extensive code modifications.

Reply:
This makes sense to me!

Reply:
Just updated!

Reply:
@cjvogl, if you agree with this pull request, could you please approve it?

Reply:
I'm not technically an assigned reviewer (listed in the table in the description), so I only took a quick look.  If you want me to do a full review, I can but it won't be for a while (I'm on travel for the next 7 days).  FWIW, you shouldn't need my approval for this to be merged, especially since it already has 2 approvals.

Reply:
Oh, I see. In that case, don't worry about it. Have a great trip :)

Reply:
Thanks! and thanks for putting this PR together!